cpp/src/strings/repeat_strings.cu

```diff
@@ -0,0 +1,200 @@
+
+namespace cudf {
+namespace strings {
+namespace detail {
+
+string_scalar repeat_strings(string_scalar const& input,
+                             size_type repeat_times,
+                             rmm::cuda_stream_view stream,
+                             rmm::mr::device_memory_resource* mr)
+{
+  if (!input.is_valid(stream)) { return string_scalar("", false, stream, mr); }
+  if (input.size() == 0 || repeat_times <= 0) { return string_scalar("", true, stream, mr); }
+  if (repeat_times == 1) { return string_scalar(input, stream, mr); }
+
+  CUDF_EXPECTS(input.size() <= std::numeric_limits<size_type>::max() / repeat_times,
+               "The output string has size that exceeds the maximum allowed size.");
+
+  auto buff =
+    rmm::device_buffer(repeat_times * input.size(), stream, rmm::mr::get_current_device_resource());
+  auto const iter     = thrust::make_counting_iterator(0);
+  auto const str_size = input.size();
+
+  // Pull data from the input string into each byte of the output string.
+  thrust::transform(rmm::exec_policy(stream),
+                    iter,
+                    iter + repeat_times * str_size,
+                    static_cast<char*>(buff.data()),
+                    [in_ptr = input.data(), str_size] __device__(const auto idx) {
+                      return in_ptr[idx % str_size];
+                    });
+
+  // TODO: using move constructor when https://github.com/rapidsai/cudf/pull/8428 merged
+  // return string_scalar(std::move(buff));
+  return string_scalar(string_view(static_cast<char*>(buff.data()), buff.size()), true, stream, mr);
+}
+
+namespace {
+/**
+ * @brief Generate a strings column in which each row is an empty or null string.
+ *
+ * The output strings column has the same bitmask as the input column.
+ */
+auto generate_empty_output(strings_column_view const& input,
+                           size_type strings_count,
+                           rmm::cuda_stream_view stream,
+                           rmm::mr::device_memory_resource* mr)
+{
+  auto chars_column = create_chars_child_column(strings_count, 0, stream, mr);
+
+  static_assert(std::is_same_v<offset_type, int32_t> && std::is_same_v<size_type, int32_t>);
+  auto offsets_column = make_numeric_column(
+    data_type{type_id::INT32}, strings_count + 1, mask_state::UNALLOCATED, stream, mr);
+  CUDA_TRY(cudaMemsetAsync(offsets_column->mutable_view().template begin<offset_type>(),
+                           0,
+                           (strings_count + 1) * sizeof(offset_type),
+                           stream.value()));
+
+  return make_strings_column(strings_count,
+                             std::move(offsets_column),
+                             std::move(chars_column),
+                             input.null_count(),
+                             input.null_count()
+                               ? cudf::detail::copy_bitmask(input.parent(), stream, mr)
+                               : rmm::device_buffer{},
```

Comment:
Unfortunately this doesn't work as expected because `copy_bitmask` does checking using `nullable()`, not `null_count()`, which fails my unit tests in the cases of sliced column (the sliced segment (of a column having nulls) does not have any null).

Reply:
You can use `CUDF_TEST_EXPECT_COLUMNS_EQUIVALENT` for those cases.

Reply:
I mean, if using `copy_bitmask` then the output column may have a bitmask even when it does not have any null (such as the sliced column example above).

I'm not sure if we should accept this result? I feel this is a waste of resources (memory + CPU) to do copying the bitmask containing all valid values. On the other hand, by (always) copying the bitmask we will have a consistency between the input and output columns: if the input column has a bitmask then the output column always has a bitmask.

Reply:
Thanks for this. I have applied your suggestion to steer the result toward "consistency" instead of "efficiency". Maybe having "consistency" is more reasonable.